benchmark/src/jmh/java/io/opentelemetry/benchmark/UsedMemoryProfiler.java

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.opentelemetry.benchmark;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import org.openjdk.jmh.*;
+import org.openjdk.jmh.annotations.*;
+import org.openjdk.jmh.infra.*;
+import org.openjdk.jmh.profile.*;
+import org.openjdk.jmh.results.*;
+
+public class UsedMemoryProfiler implements InternalProfiler {
+  private long totalHeapBefore;
+  private long usedHeapBefore;
+
+  @Override
+  public String getDescription() {
+    return "Used memory heap profiler";
+  }
+
+  @Override
+  public void beforeIteration(BenchmarkParams benchmarkParams, IterationParams iterationParams) {
+    System.gc();
+    System.runFinalization();
+
+    totalHeapBefore = Runtime.getRuntime().totalMemory();
+    usedHeapBefore = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+  }
+
+  @Override
+  public Collection<? extends Result> afterIteration(
+      BenchmarkParams benchmarkParams, IterationParams iterationParams, IterationResult result) {
+
+    long totalHeap = Runtime.getRuntime().totalMemory();
+    long usedHeap = Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory();
+
+    Collection<ScalarResult> results = new ArrayList<>();
+    results.add(
+        new ScalarResult("·heap.total.before", totalHeapBefore, "bytes", AggregationPolicy.AVG));
```

Comment:
The special character (middot?) worries me a little, but I should really stop being afraid of unicode, so let's keep it 👍

Reply:
I added it only because I saw it being used in JMH profiler's output, but I am not a huge fan of it either. Actually I think it is going to be cleaner without it